File: vrt/dotnet/VisualAssert.cs
// Visual regression helper for Playwright .NET (NUnit flavour).
//
// Dependencies (add to your .csproj):
//   <PackageReference Include="Microsoft.Playwright.NUnit" Version="1.44.0" />
//   <PackageReference Include="SixLabors.ImageSharp"       Version="3.1.0"  />
//
// Usage in a test:
//   await VisualAssert.AssertScreenshotAsync(Page, "homepage");
//
// First run creates the baseline. Subsequent runs diff against it.
// Commit the Screenshots/ directory to source control.

using System;
using System.IO;
using System.Threading.Tasks;
using Microsoft.Playwright;
using NUnit.Framework;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;

public static class VisualAssert
{
    private const string BaselineDir = "Screenshots";
    private const string DiffDir     = "TestResults/Diffs";
    private const double DefaultAllowedDiffRatio = 0.001; // 0.1 % of pixels
    private const int    ColorTolerance = 10;             // per-channel delta

    public static async Task AssertScreenshotAsync(
        IPage page,
        string name,
        double allowedDiffRatio = DefaultAllowedDiffRatio)
    {
        Directory.CreateDirectory(BaselineDir);
        Directory.CreateDirectory(DiffDir);

        var actual       = await page.ScreenshotAsync();
        var baselinePath = Path.Combine(BaselineDir, $"{name}.png");

        if (!File.Exists(baselinePath))
        {
            await File.WriteAllBytesAsync(baselinePath, actual);
            Console.WriteLine($"[VRT] Baseline created: {baselinePath}");
            return;
        }

        using var expected  = Image.Load<Rgba32>(baselinePath);
        using var actualImg = Image.Load<Rgba32>(actual);

        if (expected.Size != actualImg.Size)
            actualImg.Mutate(x => x.Resize(expected.Width, expected.Height));

        int diffPixels = 0;
        int total      = expected.Width * expected.Height;
        using var diffImg = expected.Clone();

        for (int y = 0; y < expected.Height; y++)
        {
            for (int x = 0; x < expected.Width; x++)
            {
                var e = expected[x, y];
                var a = actualImg[x, y];
                bool different =
                    Math.Abs(e.R - a.R) +
                    Math.Abs(e.G - a.G) +
                    Math.Abs(e.B - a.B) > ColorTolerance;

                if (different)
                {
                    diffPixels++;
                    diffImg[x, y] = new Rgba32(255, 0, 0, 255);
                }
                else
                {
                    diffImg[x, y] = new Rgba32(e.R, e.G, e.B, 128);
                }
            }
        }

        double ratio = (double)diffPixels / total;
        if (ratio > allowedDiffRatio)
        {
            var diffPath = Path.Combine(DiffDir, $"{name}-diff.png");
            await diffImg.SaveAsPngAsync(diffPath);
            Assert.Fail(
                $"Visual regression: {name} — {ratio:P2} pixels differ " +
                $"(limit {allowedDiffRatio:P2}). Diff: {diffPath}");
        }
    }
}

File: vrt/python/visual_assert.py
"""
Visual regression helper for pytest-playwright.

Dependencies:
    pip install pillow pytest-playwright

Usage in a test:
    from visual_assert import assert_screenshot

    def test_homepage(page):
        page.goto("https://example.com")
        assert_screenshot(page, "homepage")

First run creates the baseline. Subsequent runs diff against it.
Commit the screenshots/ directory to source control.
"""

from __future__ import annotations

import io
from pathlib import Path

import pytest
from PIL import Image, ImageChops

BASELINE_DIR = Path("tests/screenshots")
DIFF_DIR = Path("test-results/diffs")
ALLOWED_DIFF_RATIO = 0.001  # 0.1 % of pixels


def assert_screenshot(
    page,
    name: str,
    *,
    allowed_diff_ratio: float = ALLOWED_DIFF_RATIO,
) -> None:
    BASELINE_DIR.mkdir(parents=True, exist_ok=True)
    DIFF_DIR.mkdir(parents=True, exist_ok=True)

    actual_bytes = page.screenshot()
    baseline = BASELINE_DIR / f"{name}.png"

    if not baseline.exists():
        baseline.write_bytes(actual_bytes)
        print(f"\n[VRT] Baseline created: {baseline}")
        return

    expected_img = Image.open(baseline).convert("RGB")
    actual_img = Image.open(io.BytesIO(actual_bytes)).convert("RGB")

    if expected_img.size != actual_img.size:
        actual_img = actual_img.resize(expected_img.size, Image.LANCZOS)

    diff = ImageChops.difference(expected_img, actual_img)
    pixel_data = list(diff.getdata())
    differing = sum(1 for r, g, b in pixel_data if r + g + b > 0)
    ratio = differing / len(pixel_data)

    if ratio > allowed_diff_ratio:
        diff_path = DIFF_DIR / f"{name}-diff.png"
        diff.save(diff_path)
        pytest.fail(
            f"Visual regression: {name} — {ratio:.2%} pixels differ "
            f"(limit {allowed_diff_ratio:.2%}). Diff: {diff_path}"
        )
